packages/@angular/cli/upgrade/version.ts

```diff
@@ -143,6 +143,51 @@ export class Version {
     }
   }
 
+  static assertTypescriptVersion(projectRoot: string) {
+    if (!CliConfig.fromGlobal().get('warnings.typescriptMismatch')) {
+      return;
+    }
+    let compilerVersion: string, tsVersion: string;
+    try {
+      compilerVersion = requireProjectModule(projectRoot, '@angular/compiler-cli').VERSION.full;
+      tsVersion = requireProjectModule(projectRoot, 'typescript').version;
+    } catch (_) {
+      console.error(bold(red(stripIndents`
+        Versions of @angular/compiler-cli and typescript could not be determined.
+        The most common reason for this is a broken npm install.
+
+        Please make sure your package.json contains both @angular/compiler-cli and typescript in
+        devDependencies, then delete node_modules and package-lock.json (if you have one) and
+        run npm install again.
+      `)));
+      process.exit(2);
+    }
+
+    const versionCombos = [
+      { compiler: '>=2.3.1 <3.0.0', typescript: '>=2.0.2 <2.3.0' },
+      { compiler: '>=4.0.0 <5.0.0', typescript: '>=2.1.0 <2.4.0' },
+      { compiler: '>=5.0.0 <6.0.0', typescript: '>=2.4.0 <2.6.0' }
+    ];
+
+    const currentCombo = versionCombos.find((combo) => satisfies(compilerVersion, combo.compiler));
+
+    if (currentCombo && !satisfies(tsVersion, currentCombo.typescript)) {
+      // First line of warning looks weird being split in two, disable tslint for it.
+      // tslint:disable:max-line-length
```

Comment:
I don't think you have to disable max-length. The first line can be split in two (and be easier to read).